src/fparser/statements.py

```diff
@@ -1748,47 +1750,169 @@ class Case(Statement):
+        of code '''
         line = self.item.get_line()[4:].lstrip()
         if line.startswith('('):
-            i = line.find(')')
-            items = split_comma(line[1:i].strip(), self.item)
-            line = line[i+1:].lstrip()
+            idx = line.find(')')
+            items = split_comma(line[1:idx].strip(), self.item)
+            line = line[idx+1:].lstrip()
         else:
-            assert line.lower().startswith('default'),`line`
+            assert line.lower().startswith('default'), repr(line)
             items = []
             line = line[7:].lstrip()
-        for i in range(len(items)):
-            it = self.item.copy(items[i])
-            rl = []
-            for r in it.get_line().split(':'):
-                rl.append(it.apply_map(r.strip()))
-            items[i] = rl
+        for idx in range(len(items)):
+            itm = self.item.copy(items[idx])
+            rlst = []
+            for part in itm.get_line().split(':'):
+                rlst.append(itm.apply_map(part.strip()))
+            items[idx] = rlst
         self.items = items
         self.name = line
         parent_name = getattr(self.parent, 'name', '')
-        if self.name and self.name!=parent_name:
-            self.warning('expected case-construct-name %r but got %r, skipping.'\
-                         % (parent_name, self.name))
+        if self.name and self.name != parent_name:
+            self.warning("Expected case-construct-name {0} but got {1}, "
+                         "skipping.".format(parent_name, self.name))
             self.isvalid = False
         return
 
     def tofortran(self, isfix=None):
+        ''' Return the Fortran for this object as a string '''
         tab = self.get_indent_tab(isfix=isfix)
-        s = 'CASE'
+        txt = tab + 'CASE'
         if self.items:
-            l = []
+            lst = []
             for item in self.items:
-                l.append((' : '.join(item)).strip())
-            s += ' ( %s )' % (', '.join(l))
+                lst.append((' : '.join(item)).strip())
+            txt += ' ( %s )' % (', '.join(lst))
         else:
-            s += ' DEFAULT'
+            txt += ' DEFAULT'
         if self.name:
-            s += ' ' + self.name
-        return s
-    def analyze(self): return
+            txt += ' ' + self.name
+        return txt
+
+    def analyze(self):
+        return
+
+
+class TypeIs(Statement):
+    """
+    TYPE IS <type-selector> [ <case-constract-name> ]
+    <type-selector> = ( <type-value-range-list> )
+    <type-value-range> = <case-value>
+    <case-value> = <char>
+    """
+    match = re.compile(r'type\b\s*is\b\s*\(.*\)\s*\w*\Z', re.I).match
+
+    def process_item(self):
+        ''' Populate the state of this object by parsing the associated
+        line of code '''
+        line = self.item.get_line()[4:].lstrip()
+        idx = line.find('(')
+        if idx > 0:
+            line = line[idx:]
+            idx = line.find(')')
```

Comment:
Argh, another bug (in the original code). I've fixed this in 17d4219 - now the case expression is correctly treated as a sub-expression.